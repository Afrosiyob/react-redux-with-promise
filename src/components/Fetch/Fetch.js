import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import axios from "axios";

const Fetch = (props) => {
  const { url = "some url" } = props;

  const [data, setData] = useState(null);

  useEffect(() => {
    let mounted = true;

    const loadData = async () => {
      const response = await axios.get(url);

      if (mounted) {
        setData(response.data);
      }
    };

    loadData();

    return () => {
      mounted = false;
    };
  }, [url]);

  if (data) {
    return (
      <>
        <h1>Fetch component</h1>
        {data.greeting}
      </>
    );
  } else {
    return (
      <>
        <h1>Fetch component</h1>
        <p>no data</p>
      </>
    );
  }
};

Fetch.propTypes = {
  url: PropTypes.string.isRequired,
};

export default Fetch;
