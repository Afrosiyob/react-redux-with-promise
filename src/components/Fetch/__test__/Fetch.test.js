import { render, waitFor, screen, cleanup } from "@testing-library/react";
import axiosMock from "axios";
import Fetch from "../Fetch";
import renderer from "react-test-renderer";
afterEach(cleanup);

test("fetch data and display", async () => {
  axiosMock.get.mockResolvedValueOnce({
    data: {
      greeting: "hello there...",
    },
  });

  const url = "/greeting";

  const { container } = render(<Fetch url={url} />);

  expect(container).toHaveTextContent(/fetch component/i);

  const resolvedBox = await waitFor(() => container);

  expect(resolvedBox).toHaveTextContent("hello there...");

  expect(axiosMock.get).toHaveBeenCalledTimes(1);

  expect(axiosMock.get).toHaveBeenCalledWith(url);
});

test("fetch data snapshot", async () => {
  const url = "/greeting";
  const component = renderer.create(<Fetch url={url} />);
  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
