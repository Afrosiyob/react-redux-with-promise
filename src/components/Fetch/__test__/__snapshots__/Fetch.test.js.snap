// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch data snapshot 1`] = `
Array [
  <h1>
    Fetch component
  </h1>,
  <p>
    no data
  </p>,
]
`;
